### PR TITLE
NN-4066 changes required to update server, error handling and redirect path

### PR DESCRIPTION
**Pull request by timoj58**

### CODE REVIEW

File: server/routes/prisonerReport/prisonerReportPage.ts [hunk 1]
```diff
@@ -162,14 +176,39 @@ export default class prisonerReportRoutes {
...
-    return res.redirect(adjudicationUrls.allCompletedReports.root)
+    const url = req.query.referrer as string
+    const urlStatus = req.query.status as string
+    const urlToDate = req.query.toDate as string
+
+    try {
+      await this.reportedAdjudicationsService.reviewAdjudication(adjudicationNumber, status, reason, details, user)
+      return res.redirect(`${url}&toDate=${urlToDate}&status=${urlStatus}`)
+    } catch (e) {
+      if (e.status === 400) {
```
Comment (sp-watson) [inline server/routes/prisonerReport/prisonerReportPage.ts:197]: Hmm, yes I see that 400 is thrown if the user attempts an invalid transition - but is there anything they can do about it?

This is why we generally assume any API error cannot be resolved by the user, and therefore we don't tell them about it (we show a "sorry" page)

File: integration_tests/plugins/index.ts [hunk 2]
```diff
@@ -51,5 +51,6 @@ export default (on: (string, Record) => void): void => {
     stubGetOffenceRule: adjudications.stubGetOffenceRule,
     stubSaveOffenceDetails: adjudications.stubSaveOffenceDetails,
     verifySaveOffenceDetails: adjudications.verifySaveOffenceDetails,
+    stubReviewAdjudications: adjudications.stubReviewAdjudications,
```
[the remark this answers is not in the document]
Comment (timoj58) [reply integration_tests/plugins/index.ts]: name changed

File: server/data/manageAdjudicationsClient.ts [hunk 3]
```diff
@@ -86,6 +87,16 @@ export default class ManageAdjudicationsClient {
     })
   }
 
+  async reviewAdjudication(
```
[the remark this answers is not in the document]
Comment (timoj58) [reply server/data/manageAdjudicationsClient.ts]: changed

File: server/routes/prisonerReport/prisonerReportReview.test.ts [hunk 4]
```diff
@@ -14,6 +14,7 @@ jest.mock('../../services/userService.ts')
 jest.mock('../../services/reportedAdjudicationsService.ts')
```
[the remark this answers is not in the document]
Comment (timoj58) [reply server/routes/prisonerReport/prisonerReportReview.test.ts:14]: test added

File: server/routes/prisonerReport/prisonerReportPage.ts [hunk 5]
```diff
@@ -31,19 +31,28 @@ const getVariablesForPageType = (
...
   adjudicationNumber: number,
-  draftAdjudicationNumber: number
+  draftAdjudicationNumber: number,
+  req: Request
 ) => {
   if (pageOptions.isReviewerView()) {
+    let returnLinkUrl = adjudicationUrls.allCompletedReports.root
+    if (req.query.referrer !== undefined) {
+      const url = req.query.referrer as string
+      const urlStatus = req.query.status as string
```
[the remark this answers is not in the document]
Comment (timoj58) [reply server/routes/prisonerReport/prisonerReportPage.ts]: encode whole string

File: server/routes/prisonerReport/prisonerReportPage.ts (hunk 1 above)
Comment (timoj58) [reply server/routes/prisonerReport/prisonerReportPage.ts]: removed

File: server/utils/urlGenerator.ts [hunk 6]
```diff
@@ -182,7 +182,10 @@ const adjudicationUrls = {
     },
     urls: {
       report: (adjudicationNumber: number) => `${adjudicationUrls.prisonerReport.root}/${adjudicationNumber}/report`,
-      review: (adjudicationNumber: number) => `${adjudicationUrls.prisonerReport.root}/${adjudicationNumber}/review`,
+      review: (adjudicationNumber: number, originalUrl = '') =>
+        originalUrl === ''
```
[the remark this answers is not in the document]
Comment (timoj58) [reply server/utils/urlGenerator.ts]: print report also has the encoded string

**Review by sp-watson — APPROVED** [review]

**Pull request merged**